Tuple and array arguments supported in `resim`

lrubasze opened the pull request:
## Summary
Support for passing Tuple and Array values as function/method arguments in `resim` .
Examples:

- Array of Strings
```
resim call-function <package_address> <blueprint_name> <function_name> '["fizz","buzz"]'
```
- Tuple of `ResourceAddress`, `NonFungibleLocalId`
```
resim call-function <package_address> <blueprint_name> <function_name> \
 '(resource_sim1nfzf2h73frult99zd060vfcml5kncq3mxpthusm9lkglvhsr0guahy,#1#)'
```

dhedey commented on simulator/src/utils/common_instructions.rs:
```diff
@@ -171,9 +172,237 @@ macro_rules! matches_bucket {
+    let mut chars = argument.chars();
+
+    while let Some(c) = chars.next() {
+        if c.is_whitespace() {
```
I think instead we should have the check contingent on `round_brackets_level == 0 && square_brackets_level == 0`

lrubasze replied:
> This discards all whitespace inside strings contained inside tuples and arrays, I think this is a little too heavy handed, e.g. `(hello world,123)` would parse as `("helloworld", 123)`

Food point 🤦

dhedey commented on simulator/src/utils/common_instructions.rs:
```diff
@@ -171,9 +172,237 @@ macro_rules! matches_bucket {
+
+// Splits argument tuple or array elements delimited with "," into vector of elements.
+// Elements with brackets (round or square) are taken as is (even if they may include ",")
+fn split_argument_tuple_or_array(argument: &str) -> Result<Vec<String>, BuildCallArgumentError> {
```
... Although we could also consider adding string support with `"` to the below parser (and the string parser would discard start and stopping `"`) -- i.e. we could have an "is in string" flag, and avoid interpreting any other characters until we see a `"`.

lrubasze replied:
Yeah, surrounding strings with `"` would make things easier, but it would be quite different of other places in `resim`, where string could be supplied. Not sure if want it...

dhedey replied:
I guess I was thinking that it could be optional, so `abc`, `"abc"` and `"abc [ asda"` are all acceptable for a string. But it does further complicate the splitter/parser!

dhedey reviewed (commented):
This is really cool! :D 

Side thoughts:
* This is not the standard approach to implementing a parser (normally as per manifest land, you'd implement a lexer and parser) which is why there are quite a lot of edge cases. But the code's a lot more compact. I don't think it's a problem and I don't think it needs changing; but I think it could do with a few of the edge cases being neatened out, and documented a little more.
* A part of me wonders if we should accept manifest syntax or programmatic json or something (where we have proper lexers already) _but_ I think I prefer your approach tbh :)

Separately, I wonder if we should detect a well known type id of NonFungibleGlobalId, and allow that to be provided as the canonical resource_address:local_id string (as well as the tuple)?

lrubasze commented:
> Separately, I wonder if we should detect a well known type id of NonFungibleGlobalId, and allow that to be provided as the canonical resource_address:local_id string (as well as the tuple)?

This is exactly what I was thinking. In fact I've implemented it in the first [commit](https://github.com/radixdlt/radixdlt-scrypto/pull/1727/commits/e53b466903cf5cbb4253f9c5c493c9dd129d6a05) of this PR, but I replaced it with Tuple and Array approach. But I believe it would be really nice to support it as well.

lrubasze commented on simulator/src/utils/common_instructions.rs:
```diff
@@ -171,9 +172,237 @@ macro_rules! matches_bucket {
+            }
+            _ => match prev_c {
+                Some(prev_c) if prev_c == ']' || prev_c == ')' => {
+                    return Err(BuildCallArgumentError::FailedToParse(format!(
```
I think comma should be mandatory, so would prefer to fix it.
